Utilities/DataCompression/include/DataCompression/DataDeflater.h

```diff
@@ -41,132 +64,130 @@ namespace AliceO2 {
+    int nWords = 0;
+    if (mFilledBits > 0) {
+      writer(mCurrent);
+      ++nWords;
+    }
+    mCurrent = 0;
+    mFilledBits = 0;
+    return nWords;
   }
 
   /**
    * Write number of bits
    * value contains number of valid LSBs given by bitlength
    *
-   * TODO: that function might be renamed to simply 'Write' in conjunction
+   * TODO: that function might be renamed to simply 'write' in conjunction
    * with a mixin approach. Every deflater mixin instance has only one
-   * 'Write' function and does internally the necessary conversions to
-   * finally use 'Write' of the mixin base.
+   * 'write' function and does internally the necessary conversions to
+   * finally use 'write' of the mixin base.
    */
-  template <typename ValueType>
-  int WriteRaw(ValueType value, uint16_t bitlength) {
-    uint16_t bitsWritten = 0;
+  template <typename ValueType, typename WriterT>
+  int writeRaw(ValueType value, uint16_t bitlength, WriterT writer) {
+    auto bitsToWrite = bitlength;
     if (bitlength > 8*sizeof(ValueType)) {
       // TODO: error policy
-      bitlength = 8*sizeof(ValueType);
+      throw std::runtime_error("bit length exceeds width of the data type");
     }
-    while (bitsWritten < bitlength) {
-      if (mCurrent == mEnd) {
-        //break; // depending on error policy
-        return -ENOSPC;
+    while (bitsToWrite > 0) {
+      if (mFilledBits == TargetBitWidth) {
+        mFilledBits=0;
+        writer(mCurrent);
+        mCurrent = 0;
       }
-      _TargetType& current = *mCurrent;
       // write at max what is left to be written
-      uint16_t writeNow = bitlength - bitsWritten;
-      ValueType mask = 1<<(writeNow); mask -= 1;
+      auto writeNow = bitsToWrite;
       // write one element of the target buffer at a time
-      if (writeNow > 8*sizeof(_TargetType)) writeNow = 8*sizeof(_TargetType);
+      if (writeNow > TargetBitWidth) writeNow = TargetBitWidth;
       // write the remaining space in the current element
-      uint16_t capacity=8*sizeof(_TargetType)-mBitPosition;
+      auto capacity = TargetBitWidth - mFilledBits;
       if (writeNow > capacity) writeNow = capacity;
-      ValueType activebits=(value&mask)>>(bitlength-bitsWritten-writeNow);
-      activebits<<(capacity-writeNow);
-      mBitPosition+=writeNow;
-      bitsWritten+=writeNow;
-      if (mBitPosition==8*sizeof(_TargetType)) {
-        mBitPosition=0;
-        mCurrent++;
-      } // pedantic check: should never exceed the taget type size
+      auto mask = (((ValueType)1 << writeNow) - 1) << (bitsToWrite - writeNow);
+      auto activebits = (value&mask) >> (bitsToWrite - writeNow);
+      mCurrent |= activebits << (capacity-writeNow);
+      mFilledBits += writeNow;
+      bitsToWrite -= writeNow;
+      assert(mFilledBits <= TargetBitWidth);
     }
-    return bitsWritten;
+    return bitlength - bitsToWrite;
   }
 
-  int WriteRaw(bool bit) {
-    return WriteRaw(bit, 1);
+  int writeRaw(bool bit) {
```

Comment:
good spot, the function does not compile any more because the scheme has been changed, but it was not used in the test, so this was undetected. Deleted.